src/main/java/app/linguistai/bmvp/service/gamification/quest/QuestService.java

```diff
@@ -0,0 +1,334 @@
+                this.processUseWordQuest(quest, criteria, (QQuestSendMessage) action);
+            }
+            else if (type instanceof SendMessageCriteria criteria) {
+                this.processSendMessageQuest(quest, (QQuestSendMessage) action);
+            }
+        }
+        catch (Exception e) {
+            log.error("Process quest type action failed for email {}", user.getEmail(), e);
+            throw new SomethingWentWrongException();
+        }
+    }
+
+    @Transactional
+    private void checkQuestResult(Quest quest) throws Exception {
+        if (
+            quest.getProgress().getTimes() < quest.getCompletionCriteria().getTimes()
+            || quest.isCompleted()
+        ) {
+            return;
+        }
+
+        // If we are here, we know the target quest times is achieved
+        // and the quest is not marked as complete,
+        // so we award the reward to the user
+        xpService.awardQuestReward(quest.getUser().getEmail(), quest.getReward());
+
+        // And mark the quest as completed
+        quest.setCompleted(true);
+        questRepository.save(quest);
+    }
+
+    @Transactional
+    private void processSendMessageQuest(Quest quest, QQuestSendMessage action) throws Exception {
+        Integer previousProgress = quest.getProgress().getTimes();
+
+        if (previousProgress >= quest.getCompletionCriteria().getTimes()) {
+            return;
+        }
+
+        quest.setProgress(new QuestProgress(previousProgress + 1));
+        Quest saved = questRepository.save(quest);
+
+        // If quest is completed, award the result
+        this.checkQuestResult(saved);
+    }
+
+    @Transactional
+    private void processUseWordQuest(Quest quest, UseWordCriteria criteria, QQuestSendMessage action) throws Exception {
+        Integer previousProgress = quest.getProgress().getTimes();
+
+        if (previousProgress >= quest.getCompletionCriteria().getTimes()) {
+            return;
+        }
+
+        // Check if message contains required word (case-insensitive)
+        // TR_TAG is used since most users will be in Turkey
+        // (for addressing issues with specific characters like i, I, etc.)
+        String requiredWord = criteria.getWord().toLowerCase(Locale.forLanguageTag(TR_TAG));
+        String message = action.getMessage().toLowerCase(Locale.forLanguageTag(TR_TAG));
+
+        if (!message.contains(requiredWord)) {
+            return;
+        }
+
+        quest.setProgress(new QuestProgress(previousProgress + 1));
+        Quest saved = questRepository.save(quest);
+
+        // If quest is completed, award the result
+        this.checkQuestResult(saved);
+    }
+
+    @Override
+    @Transactional
+    public void assignQuests(String email) throws Exception {
+        try {
+            // Step 1. Check if user exists
+            User user = accountRepository.findUserByEmail(email)
+                    .orElseThrow(() -> new NotFoundException("User not found"));
```

Comment:
To NotFoundException, you can give User.class.getSimpleName() as a parameter just so that it follows a similar structure to exceptions in other service classes. But this works fine too, your call

Reply:
I saw this implenentation in some classes too, but since User.getSimpleClassName won't be used in the catch block, it will be redundant I believe